### PR TITLE
[Sparse & HICache]: Enables hierarchical sparse KV cache management and scheduling for DeepSeek V32.

### PR DESCRIPTION
**Pull request by hzh0425**
## Motivation

Previously, Sparse Attention was introduced in DeepSeekV32 Mode, which selects only the top-2048 tokens to participate in attention computation, significantly improving inference performance in long-context scenarios.

However, this model still requires **caching the full KV cache in GPU memory,** meaning the storage footprint of the KV cache is not reduced. This limits the number of concurrent long-context requests the system can handle.

Therefore, we aim to integrate HICache to implement a hierarchical DSA—where the full KV cache is stored in CPU or remote store, **and only 2,048 tokens per request are kept in GPU memory**. This significantly increases the batch size during the decode phase and improves overall decode throughput.

CoAuthors:  @huangtingwei9988  @xiezhq-hermann @LingYeAI and hicache team
Upstream Issue：https://github.com/sgl-project/sglang/issues/12826
**This is the upstream branch, and we will split it into multiple smaller PRs to move forward. （**ready now**）**
Todo PR:
- [x] Define Sparse Interface: https://github.com/sgl-project/sglang/pull/14741
- [x] Seperate Index_k pool and kvcache pool https://github.com/sgl-project/sglang/pull/15807
- [x] Adding SparseCoodinator and attention backend adaptor https://github.com/sgl-project/sglang/pull/16086
- [x] Support sparse diff kenrel  & Support IO Transfer Logic https://github.com/sgl-project/sglang/pull/16984
- [x] Support DeepSeek DSA & GLM5  https://github.com/sgl-project/sglang/pull/20343
- [ ] Support two batch overlap

## Modifications

1. We have **integrated HICache to build a unified hierarchical sparse framework** that supports various sparse attention algorithms—such as [Quest](https://arxiv.org/pdf/2406.10774) and [ClusterKV](https://arxiv.org/pdf/2412.03213)—for hierarchical, sparsity-aware KV cache management.
2. Implementation of Hierarchical DSA:

      - Decoupled the DSA indexer from KV cache allocation and the req_to_token_pool mechanism: the full indexer remains on the GPU.
      - During the decode phase, each request is allocated a fixed token space of 2,048 tokens in GPU memory.
      - Before each attention layer begins, a diff kernel identifies the missing KV cache entries in GPU memory, and HICache incrementally fetches only those missing portions from CPU memory. Our observations show that, on average, only about 20% of the KV cache needs to be fetched incrementally per layer.

image

## Accuracy Tests

We evaluated several LongBench tests, and the accuracy is nearly identical to that of the original DSA version.
image

## Benchmarking and Profiling

1. Comparison Setup:
    
      - Model: dsv32 on H20 (140 GB)
      - Configuration:8DP, disabled radix tree、overlap scheduling, Average sequence length: 16K
      - Native Mode: Maximum supported batch size ≈ (249,472 / 16,384) × 8 ≈ 120
      - Hierarchical Mode: Maximum supported batch size ≈ (249,472 / 2,048) × 8 ≈ 900; Host memory usage per rank: ~72 GB
      - Decode Throughput Comparison is as follows: (**The current performance is not yet optimal**. We will continue refining and optimizing, using computation-communication overlap to hide the latency introduced by additional I/O.)

image

2. Below is an analysis of GPU memory usage in a long-text generation scenario with 4K input tokens, 16 output tokens, and 12 concurrent requests. As shown, the hierarchical DSA maintains GPU memory utilization at only around 10%, whereas the normal mode quickly exhausts nearly all available GPU memory.
image

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/mem_cache/sparsity/algorithms/deepseek_nsa.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+class DeepSeekNSAAlgorithm(BaseSparseAlgorithm):
+    """Sparse attention algorithm for DeepSeek NSA."""
+
+    def __init__(self, config, device: torch.device, **kwargs):
+        super().__init__(config, device, **kwargs)
+        self.index_topk = getattr(config, "index_topk", 2048)
+        logger.info(f"DeepSeekNSAAlgorithm initialized: index_topk={self.index_topk}")
+
+    def get_sparse_mode(self) -> SparseMode:
+        return SparseMode.ORIGINAL_WISE
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/sparsity/algorithms/deepseek_nsa.py]: let's use a clearer name for it, if it's block wise then it's block wise, or specifically dpsk-block-wise

File: python/sglang/srt/mem_cache/sparsity/algorithms/base_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+
+class SparseMode(Enum):
+    """Sparse attention granularity mode."""
+
+    PAGE_WISE = "page_wise"
+    TOKEN_WISE = "token_wise"
+    ORIGINAL_WISE = "original_wise"
+
+
+class BaseSparseAlgorithm(ABC):
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/sparsity/algorithms/base_algorithm.py:10]: let's aim for to get the algorithm part under `/mem_cache/sparsity/algorithm` as a separated PR to get merged first, with a more detailed review, proper explanation and testing

File: python/sglang/srt/mem_cache/sparsity/algorithms/base_algorithm.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+            req_to_token: Mapping tensor
+            last_extracted_tokens: [max_pool_size] global state
+
+        Returns:
+            success_mask: [bs] bool, which requests successfully updated
+        """
+        return torch.zeros_like(req_pool_indices, dtype=torch.bool)
+
+
+class PageMeanPoolingAlgorithm(BaseSparseAlgorithm):
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/sparsity/algorithms/base_algorithm.py]: is this generic to all page-wise algorithms?

File: python/sglang/srt/mem_cache/sparsity/core/representation_pool.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        end_layer: int,
+        device: torch.device,
+    ):
+        self.total_num_pages = total_num_pages
+        self.start_layer = start_layer
+        self.end_layer = end_layer
+        self.device = device
+        self.storages: Dict[str, Dict[int, torch.Tensor]] = {}
+
+    def register_storage(self, name: str, shape: tuple, dtype: torch.dtype) -> None:
```
Comment (xiezhq-hermann) [inline python/sglang/srt/mem_cache/sparsity/core/representation_pool.py]: let's not to use the term `storage` here to avoid confusion with storage as memory tiers.

**Comment by hzh0425** [comment]
> hi, may I ask is there a plan to support ds3.2 L2 hicache in PD unified-mode?

Yes; we are supporting this feature and expect to release it as soon as possible
